sklearn/decomposition/_nmf.py

```diff
@@ -803,35 +914,39 @@ def _fit_multiplicative_update(X, W, H, beta_loss='frobenius',
     previous_error = error_at_init
 
     H_sum, HHt, XHt = None, None, None
-    for n_iter in range(1, max_iter + 1):
-        # update W
-        # H_sum, HHt and XHt are saved and reused if not update_H
-        delta_W, H_sum, HHt, XHt = _multiplicative_update_w(
-            X, W, H, beta_loss, l1_reg_W, l2_reg_W, gamma,
-            H_sum, HHt, XHt, update_H)
-        W *= delta_W
-
-        # necessary for stability with beta_loss < 1
-        if beta_loss < 1:
-            W[W < np.finfo(np.float64).eps] = 0.
-
-        # update H
-        if update_H:
-            delta_H = _multiplicative_update_h(X, W, H, beta_loss, l1_reg_H,
-                                               l2_reg_H, gamma)
-            H *= delta_H
-
-            # These values will be recomputed since H changed
-            H_sum, HHt, XHt = None, None, None
 
+    for n_iter in range(0, max_iter):
+        for iter_offset, slice in enumerate(
+            gen_batches(n=n_samples, batch_size=batch_size)
+        ):
+            # update W
+            # H_sum, HHt and XHt are saved and reused if not update_H
+            delta_W, H_sum, HHt, XHt = _multiplicative_update_w(
+                X[slice], W[slice], H, beta_loss, l1_reg_W, l2_reg_W,
+                gamma, H_sum, HHt, XHt, update_H)
+            W[slice] *= delta_W
             # necessary for stability with beta_loss < 1
-            if beta_loss <= 1:
-                H[H < np.finfo(np.float64).eps] = 0.
+            if beta_loss < 1:
+                W[slice][W[slice] < np.finfo(np.float64).eps] = 0.
 
-        # test convergence criterion every 10 iterations
-        if tol > 0 and n_iter % 10 == 0:
-            error = _beta_divergence(X, W, H, beta_loss, square_root=True)
+            # update H
+            if update_H:
+                H, A, B = _multiplicative_update_h(
+                    X[slice], W[slice], H, A, B, beta_loss,
+                    l1_reg_H, l2_reg_H, gamma, rho)
+
+                # These values will be recomputed since H changed
+                H_sum, HHt, XHt = None, None, None
+
+                # necessary for stability with beta_loss < 1
+                if beta_loss <= 1:
+                    H[H < np.finfo(np.float64).eps] = 0.
+
+            iter_offset += 1
 
+        # test convergence criterion every iteration
+        if tol > 0:
+            error = _beta_divergence(X, W, H, beta_loss, square_root=True)
```

Comment:
This stopping criterion is not suited for the online nmf. If we put it inside the minibatches loop, it requires computing the loss on the full dataset at each step which is too costly. If we put it inside the max_iter loop, we only check for convergence at each epoch which can be too infrequently.

I think we need to keep this criterion here only for the normal nmf and add another stopping criterion (to be decided) inside the minibatch loop for the online mode only.

Potential stopping criterions are on ||H_new - H_old|| or on a smoothed estimation of the objective function

Reply:
I have reverted the test to the current upstream version for now.

Reply:
So if I understand correctly, avoiding the nested loop as you proposed above will also solve this problem...

Reply:
Well it makes it easier to break out of the loop, but we still need to find a good stopping criterion :)
(in the ref paper it's based on ||H_new - H_old||)